File: app/src/main/java/com/lskycity/androidtools/NetworkFragment.java
package com.lskycity.androidtools;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.DhcpInfo;
import android.net.NetworkInfo;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.text.format.Formatter;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.lskycity.support.utils.DeviceUtils;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.ArrayList;
import java.util.Enumeration;

import static android.content.Context.WIFI_SERVICE;

/**
 * Created by zhaofliu on 10/1/16.
 *
 */

public class NetworkFragment extends Fragment {

    private ArrayList<InfoBin> infoBinArrayList = new ArrayList<>();

    private NetAdapter adapter;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.activity_network_info, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ListView listView = (ListView) view.findViewById(R.id.list_view);
        listView.setAdapter(adapter = new NetAdapter(getActivity()));
        listView.setEmptyView(view.findViewById(R.id.empty_view));
        updateNetworkInfo();
        setHasOptionsMenu(true);
    }

    private void updateNetworkInfo() {

        NetworkInfo networkInfo = getActiveNetworkInfo();
        infoBinArrayList.clear();
        if(networkInfo != null) {
            infoBinArrayList.add(getNetworkType(networkInfo));
            infoBinArrayList.add(getNetworkState(networkInfo));
            infoBinArrayList.add(getNetworkExtra(networkInfo));
            infoBinArrayList.add(getNetworkExtraInfo(networkInfo));
            infoBinArrayList.addAll(getNetworkIpAddressFormInterface());
            if(networkInfo.getType() == ConnectivityManager.TYPE_WIFI) {
                infoBinArrayList.addAll(getNetworkIpAddressFormWifiManager(getActivity()));
            }
            infoBinArrayList.add(getDNSInfo());
        }
        adapter.notifyDataSetChanged();

    }

    private InfoBin getNetworkType(NetworkInfo networkInfo) {
        InfoBin bin = new InfoBin();
        bin.name = getString(R.string.network_type);
        bin.value = networkInfo.getTypeName();
        if(!TextUtils.isEmpty(networkInfo.getSubtypeName())) {
            bin.value += "["+networkInfo.getSubtypeName()+"]";
        }
        return bin;
    }

    private InfoBin getNetworkState(NetworkInfo networkInfo) {
        InfoBin bin = new InfoBin();
        bin.name = getString(R.string.network_state);
        bin.value = networkInfo.getState()+"/"+networkInfo.getDetailedState();
        return bin;
    }

    private InfoBin getNetworkExtra(NetworkInfo networkInfo) {
        InfoBin bin = new InfoBin();
        bin.name = getString(R.string.network_name);
        bin.value = networkInfo.getExtraInfo();
        return bin;
    }

    private InfoBin getNetworkExtraInfo(NetworkInfo networkInfo) {
        InfoBin bin = new InfoBin();
        bin.name = getString(R.string.network_extra_info);

        StringBuilder builder = new StringBuilder();
        if(networkInfo.isFailover()) {
            builder.append("failover, ");
        } else {
            builder.append("not failover, ");
        }

        if(networkInfo.isAvailable()) {
            builder.append("available, ");
        } else {
            builder.append("not available, ");
        }

        if(networkInfo.isRoaming()) {
            builder.append("roaming");
        } else {
            builder.append("not roaming");
        }

        bin.value = builder.toString();

        return bin;
    }

    private ArrayList<InfoBin> getNetworkIpAddressFormInterface() {

        ArrayList<InfoBin> list = new ArrayList<>(5);

        try {
            Enumeration<NetworkInterface> networkInterfaceList = NetworkInterface.getNetworkInterfaces();
            while (networkInterfaceList.hasMoreElements()) {
                NetworkInterface interfacess = networkInterfaceList.nextElement();
                String name = interfacess.getName();

                InfoBin infoBin = new InfoBin();
                infoBin.name = getString(R.string.network_interface)+"/"+name;
                infoBin.value = "";


                Enumeration<InetAddress> inetAddresses = interfacess.getInetAddresses();

                while (inetAddresses.hasMoreElements()) {
                    InetAddress address = inetAddresses.nextElement();
                    if(!address.isLoopbackAddress()) {

                        if(!TextUtils.isEmpty(infoBin.value)) {
                            infoBin.value += '\n';
                        }

                        String addressStr = address.getHostAddress();

                        int index;
                        if((index = addressStr.indexOf('%'))>0) {
                            addressStr = addressStr.substring(0, index);
                        }

                        infoBin.value += addressStr;
                    }
                }

                if(!TextUtils.isEmpty(infoBin.value)) {
                    list.add(infoBin);

                }


            }

        } catch (SocketException e) {
            e.printStackTrace();
        }

        return list;
    }

    private ArrayList<InfoBin> getNetworkIpAddressFormWifiManager(Context context) {

        ArrayList<InfoBin> list = new ArrayList<>(5);

        WifiManager wm = (WifiManager) context.getSystemService(WIFI_SERVICE);
        String ip = Formatter.formatIpAddress(wm.getConnectionInfo().getIpAddress());

        DhcpInfo dhcpInfo = wm.getDhcpInfo();

        InfoBin ipBin = new InfoBin();
        ipBin.name = getString(R.string.wifi_ip_address);
        if(DeviceUtils.isInArc()) {
            ipBin.name += "("+getString(R.string.chromebook) +")";
        }
        ipBin.value = Formatter.formatIpAddress(dhcpInfo.ipAddress) +"/"+ip;
        list.add(ipBin);

        InfoBin gateway = new InfoBin();
        gateway.name = getString(R.string.gateway);
        gateway.value = Formatter.formatIpAddress(dhcpInfo.gateway);
        list.add(gateway);

        InfoBin netmask = new InfoBin();
        netmask.name = getString(R.string.netmask);
        netmask.value = Formatter.formatIpAddress(dhcpInfo.netmask);
        list.add(netmask);

        InfoBin dns1 = new InfoBin();
        dns1.name = getString(R.string.dns1);
        dns1.value = Formatter.formatIpAddress(dhcpInfo.dns1);
        list.add(dns1);

        InfoBin dns2 = new InfoBin();
        dns2.name = getString(R.string.dns2);
        dns2.value = Formatter.formatIpAddress(dhcpInfo.dns2);
        list.add(dns2);

        InfoBin serverAddress = new InfoBin();
        serverAddress.name = getString(R.string.server_address);
        serverAddress.value = Formatter.formatIpAddress(dhcpInfo.serverAddress);
        list.add(serverAddress);

        InfoBin leaseDuration = new InfoBin();
        leaseDuration.name = getString(R.string.lease_duration);
        leaseDuration.value = Formatter.formatIpAddress(dhcpInfo.leaseDuration);
        list.add(leaseDuration);

        return list;
    }

    private  InfoBin getDNSInfo() {
        InfoBin bin = new InfoBin();

        bin.name = getString(R.string.dns);
        Class<?> SystemProperties = null;
        try {
            SystemProperties = Class.forName("android.os.SystemProperties");
            Method method = SystemProperties.getMethod("get", new Class[] { String.class });
            ArrayList<String> servers = new ArrayList<String>();
            for (String name : new String[] { "net.dns1", "net.dns2", "net.dns3", "net.dns4", }) {
                String value = (String) method.invoke(null, name);
                if (value != null && !"".equals(value) && !servers.contains(value))
                    servers.add(value);
            }

            bin.value = servers.toString();

        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        }


        return bin;
    }



    private NetworkInfo getActiveNetworkInfo() {
        ConnectivityManager connMgr = (ConnectivityManager)
                getActivity().getSystemService(Context.CONNECTIVITY_SERVICE);
        return connMgr.getActiveNetworkInfo();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.refresh, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item.getItemId() == R.id.refresh) {
            Toast.makeText(getActivity(), R.string.refreshing, Toast.LENGTH_LONG).show();
            updateNetworkInfo();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    class NetAdapter extends BaseAdapter {

        private LayoutInflater inflater;


        NetAdapter(Context context) {
            inflater = LayoutInflater.from(context);
        }


        @Override
        public int getCount() {
            return infoBinArrayList.size();
        }

        @Override
        public InfoBin getItem(int position) {
            return infoBinArrayList.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            if(convertView == null) {
                convertView = inflater.inflate(R.layout.item_config_list, parent, false);
            }
            TextView titleView = (TextView) convertView.findViewById(R.id.title);
            TextView summaryView = (TextView) convertView.findViewById(R.id.summary);
            titleView.setText(getItem(position).name);
            summaryView.setText(getItem(position).value);
            return convertView;
        }
    }

}

File: app/src/main/java/com/lskycity/androidtools/NetworkFragment.kt
package com.lskycity.androidtools

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.net.wifi.WifiManager
import android.os.Bundle
import android.text.TextUtils
import android.text.format.Formatter
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast

import com.lskycity.support.utils.DeviceUtils

import java.lang.reflect.InvocationTargetException
import java.net.NetworkInterface
import java.net.SocketException
import java.util.ArrayList

import android.content.Context.WIFI_SERVICE
import androidx.fragment.app.Fragment
import com.lskycity.support.utils.ViewUtils

/**
 * Created by zhaofliu on 10/1/16.
 *
 */

class NetworkFragment : Fragment() {

    private val infoBinArrayList = ArrayList<InfoBin>()

    private lateinit var adapter: NetAdapter



    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.activity_network_info, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val listView = view.findViewById<View>(R.id.list_view) as ListView
        adapter = NetAdapter(activity!!)
        listView.adapter = adapter
        listView.emptyView = view.findViewById(R.id.empty_view)
        updateNetworkInfo()
        setHasOptionsMenu(true)
    }

    private fun updateNetworkInfo() {

        val connMgr = activity!!.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = connMgr.activeNetworkInfo
        infoBinArrayList.clear()
        if (networkInfo != null) {
            infoBinArrayList.add(getNetworkType(networkInfo))
            infoBinArrayList.add(getNetworkState(networkInfo))
            infoBinArrayList.add(getNetworkExtra(networkInfo))
            infoBinArrayList.add(getNetworkExtraInfo(networkInfo))
            infoBinArrayList.addAll(getNetworkIpAddressFormInterface())
            if (networkInfo.type == ConnectivityManager.TYPE_WIFI) {
                infoBinArrayList.addAll(getNetworkIpAddressFormWifiManager(activity!!))
            }
            infoBinArrayList.add(getDnsInfo())
        }
        adapter.notifyDataSetChanged()

    }

    private fun getDnsInfo(): InfoBin {
        val bin = InfoBin()

        bin.name = getString(R.string.dns)
        var SystemProperties: Class<*>? = null
        try {
            SystemProperties = Class.forName("android.os.SystemProperties")
            val method = SystemProperties!!.getMethod("get", *arrayOf<Class<*>>(String::class.java))
            val servers = ArrayList<String>()
            for (name in arrayOf("net.dns1", "net.dns2", "net.dns3", "net.dns4")) {
                val value = method.invoke(null, name) as String
                if ("" != value && !servers.contains(value))
                    servers.add(value)
            }

            bin.value = servers.toString()

        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: Exception) {
            e.printStackTrace()
        }


        return bin
    }

    private fun getNetworkIpAddressFormInterface(): ArrayList<InfoBin> {

        val list = ArrayList<InfoBin>(5)

        try {
            val networkInterfaceList = NetworkInterface.getNetworkInterfaces()
            while (networkInterfaceList.hasMoreElements()) {
                val interfacess = networkInterfaceList.nextElement()
                val name = interfacess.name

                val infoBin = InfoBin()
                infoBin.name = getString(R.string.network_interface) + "/" + name
                infoBin.value = ""


                val inetAddresses = interfacess.inetAddresses

                while (inetAddresses.hasMoreElements()) {
                    val address = inetAddresses.nextElement()
                    if (!address.isLoopbackAddress) {

                        if (!TextUtils.isEmpty(infoBin.value)) {
                            infoBin.value += '\n'.toString()
                        }

                        var addressStr = address.hostAddress

                        val index: Int = addressStr.indexOf('%')
                        if (index > 0) {
                            addressStr = addressStr.substring(0, index)
                        }

                        infoBin.value += addressStr
                    }
                }

                if (!TextUtils.isEmpty(infoBin.value)) {
                    list.add(infoBin)

                }


            }

        } catch (e: SocketException) {
            e.printStackTrace()
        }

        return list
    }

    private fun getNetworkType(networkInfo: NetworkInfo): InfoBin {
        val bin = InfoBin()
        bin.name = getString(R.string.network_type)
        bin.value = networkInfo.typeName
        if (!TextUtils.isEmpty(networkInfo.subtypeName)) {
            bin.value += "[" + networkInfo.subtypeName + "]"
        }
        return bin
    }

    private fun getNetworkState(networkInfo: NetworkInfo): InfoBin {
        val bin = InfoBin()
        bin.name = getString(R.string.network_state)
        bin.value = networkInfo.state.toString() + "/" + networkInfo.detailedState
        return bin
    }

    private fun getNetworkExtra(networkInfo: NetworkInfo): InfoBin {
        val bin = InfoBin()
        bin.name = getString(R.string.network_name)
        bin.value = networkInfo.extraInfo
        return bin
    }

    private fun getNetworkExtraInfo(networkInfo: NetworkInfo): InfoBin {
        val bin = InfoBin()
        bin.name = getString(R.string.network_extra_info)

        val builder = StringBuilder()
        if (networkInfo.isFailover) {
            builder.append("failover, ")
        } else {
            builder.append("not failover, ")
        }

        if (networkInfo.isAvailable) {
            builder.append("available, ")
        } else {
            builder.append("not available, ")
        }

        if (networkInfo.isRoaming) {
            builder.append("roaming")
        } else {
            builder.append("not roaming")
        }

        bin.value = builder.toString()

        return bin
    }

    private fun getNetworkIpAddressFormWifiManager(context: Context): ArrayList<InfoBin> {

        val list = ArrayList<InfoBin>(7)

        val wm = context.applicationContext.getSystemService(WIFI_SERVICE) as WifiManager
        val ip = Formatter.formatIpAddress(wm.connectionInfo.ipAddress)

        val dhcpInfo = wm.dhcpInfo

        var name = getString(R.string.wifi_ip_address)
        if (DeviceUtils.isInArc()) {
            name = name + "(" + getString(R.string.chromebook) + ")"
        }
        val value = Formatter.formatIpAddress(dhcpInfo.ipAddress) + "/" + ip
        list.add(InfoBin(name, value))

        val gateway = InfoBin(getString(R.string.gateway), Formatter.formatIpAddress(dhcpInfo.gateway))
        list.add(gateway)

        val netmask = InfoBin(getString(R.string.netmask), Formatter.formatIpAddress(dhcpInfo.netmask))
        list.add(netmask)

        val dns1 = InfoBin(getString(R.string.dns1), Formatter.formatIpAddress(dhcpInfo.dns1))
        list.add(dns1)

        val dns2 = InfoBin(getString(R.string.dns2), Formatter.formatIpAddress(dhcpInfo.dns2))
        list.add(dns2)

        val serverAddress = InfoBin(getString(R.string.server_address), Formatter.formatIpAddress(dhcpInfo.serverAddress))
        list.add(serverAddress)

        val leaseDuration = InfoBin(getString(R.string.lease_duration), Formatter.formatIpAddress(dhcpInfo.leaseDuration))
        list.add(leaseDuration)

        return list
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.refresh, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item!!.itemId == R.id.refresh) {
            Toast.makeText(activity, R.string.refreshing, Toast.LENGTH_LONG).show()
            updateNetworkInfo()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    internal inner class NetAdapter(context: Context) : BaseAdapter() {

        override fun getCount(): Int = infoBinArrayList.size

        override fun getItem(position: Int): InfoBin = infoBinArrayList[position]

        override fun getItemId(position: Int): Long = position.toLong()

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val rootView = convertView ?: ViewUtils.inflate(R.layout.item_config_list, parent, false);

            val titleView = rootView.findViewById<View>(R.id.title) as TextView
            val summaryView = rootView.findViewById<View>(R.id.summary) as TextView
            titleView.text = getItem(position).name
            summaryView.text = getItem(position).value
            return rootView
        }
    }

}
